### DISCUSSION TITLE
How to enable the View Source of the Page functionality / button

### DISCUSSION THREAD

**Discussion by lukaszgiez** [squidfunk/mkdocs-material#4055] [Q&A] [answered]
Hello,

For some time now, I have noticed the new functionality on Material for Mkdocs tutorial pages: the **View Source of the Page** button, which then displays markdown code. How can I enable it on my markdown pages in Material?

**Comment by squidfunk** [comment, marked as answer]
__Update__: [this functionality is available as of 9.0](https://squidfunk.github.io/mkdocs-material/setup/adding-a-git-repository/#code-actions).

---

This is a custom override. I have no plans adding this functionality, because it is solely useful for projects that need you to inspect the Markdown of the page, and there are only a handful of them. Here's the override:

https://github.com/squidfunk/mkdocs-material/blob/2892706b70036b42e3e97fcbfd3198287c998456/src/overrides/partials/content.html#L23-L41

You can follow the [customization guide](https://squidfunk.github.io/mkdocs-material/customization/) to apply it yourself.

**Reply by squidfunk** [reply]
It's also heavily dependent on GitHub, which means it doesn't generalize to GitLab or other source providers. Thus, in general, not a good fit to include this generally. However, overriding is simple.

**Reply by Juke34** [reply]
First thank you so much for this Wonderfull work.
There is apparently a problem of interoperability:
  * the `edit_url` use `master` as default branch while new repo use `main`  
  * the `edit_url` use `docs` as default docs_dir. But when changing it in the mkdocs.yml e.g. `docs_dir: pages/` this is not reflected within the  `edit_url`.

So we get this url
```
https://raw.githubusercontent.com/user/repo/main/pages/file.md
```
instead of 
```
https://raw.githubusercontent.com/user/repo/master/docs/file.md
```

Do you know any trick to fix that limitation?

**Reply by squidfunk** [reply]
Are you using the override from above? If so, please don't. This functionality has been integrated, [see our documentation](https://squidfunk.github.io/mkdocs-material/setup/adding-a-git-repository/#code-actions).

**Reply by Juke34** [reply]
I have tried both. As I had the problematic behaviour I described above with the new version (via   content.action.view and content.code.annotate in the yml file), I then tried via the overrides approach, which was also not successful neither.

**Reply by squidfunk** [reply]
Without knowing exactly how you configured Material for MkDocs, we can only guess (as I just did). You can provide a minimal reproduction, we can then look into it. For this, please consult our [reproduction guide](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/).

**Reply by squidfunk** [reply]
Additionally, please consult MkDocs documentation on [`edit_uri`](https://www.mkdocs.org/user-guide/configuration/#edit_uri) – it's very flexible.

**Reply by Juke34** [reply]
Thank you for your help.
Would the access to my current project be sufficient [here](https://juke34.github.io/workshop-reproducible-research/home/)?
I didn't know about `edit_uri`, it sounds perfect to find a way to fix the url. I will give a try.

**Reply by Juke34** [reply]
Great `edit_uri` works perfectly! Thank again @squidfunk for your kind help.
If any other have the same problem () here is the way to go:

add `edit_uri: edit/main/pages/`  in mkdocs.yml where `main` is the branch in the repo and `pages`  is the value `docs_dir: pages/` set in the mkdocs.yml . If you didn't use the docs_dir parameter you must use `docs` instead .

**Comment by samzong** [comment]
this a good desgin for me.

**Comment by gledos** [comment]
Very useful feature, hopefully it will be mentioned directly in the https://squidfunk.github.io/mkdocs-material/setup/adding-a-git-repository/ page.

**Comment by squidfunk** [comment]
The view source functionality was added in [`28d64a289`](https://github.com/squidfunk/mkdocs-material/commit/28d64a289a79fc21b482c2bdd47201a49cec930f#diff-d1850a9d3d9dc4dc5eb2a685778677b763d2f832c9349cc185bf0303e70e414c). It's part of Material for MkDocs 9, which is tracked in #4714. The edit and view source button now need to be enabled explicitly:

``` yaml
theme:
  features:
    - content.action.edit
    - content.action.view
```

**Reply by nothingface0** [reply]
Thanks, I was wondering what happened when I upgraded from version 8!

**Reply by squidfunk** [reply]
It's all documented in our [upgrade guide](https://squidfunk.github.io/mkdocs-material/upgrade/#upgrading-from-8x-to-9x).

**Comment by abhibeckert** [comment]
> it is solely useful for projects that need you to inspect the Markdown of the page, and there are only a handful of them

I don't agree with that. One of the main reasons markdown is popular is because the source is human readable, and therefore it should be possible to read it (otherwise, why use Markdown?)

Github isn't the only project management system with a web interface. Almost all of them have that feature.

> It's all documented in our [upgrade guide](https://squidfunk.github.io/mkdocs-material/upgrade/#upgrading-from-8x-to-9x).

It's mentioned there, but not properly documented. I haven't been able to get it to work yet (I'm sure I'm just missing something obvious, but I'll figure it out).

**Comment by XobSod** [comment]
@abhibeckert 
Did you manage to get it to work?
Simply adding
``` yaml
theme:
  features:
    - content.action.edit
    - content.action.view
```
does nothing.
I am on:
``` sh
»  pip show mkdocs-material
Name: mkdocs-material
Version: 9.1.18
```

**Reply by niclasheinz** [reply]
There is the Documentation: https://squidfunk.github.io/mkdocs-material/setup/adding-a-git-repository/?h=edit+url#code-actions
